server/api/shifts/getAllShifts.php

```diff
@@ -0,0 +1,52 @@
+<?php
+
+$root = realpath($_SERVER["DOCUMENT_ROOT"]);
+require_once "$root/server/config.php";
+
+getAllShifts($_GET);
+
+/*
+ * The fields that CAN be returned are: shiftId, startTime, endTime, archived, createdAt, lastModifiedDate, siteId, createdBy, lastModifiedBy
+ *
+ * The data can be used to narrow down what is selected from the database:
+ * {
+ *   "shiftId": false,
+ *   "endTime": true,
+ *   "siteId": true
+ * }
+ *
+ * If the database field is missing it is assumed that you DON'T want it.
+ */
+function getAllShifts($data) {
+	GLOBAL $DB_CONN;
+	$defaultSelectColumns = array('shiftId', 'startTime', 'endTime', 'archived', 'createdAt', 'lastModifiedDate', 'siteId', 'createdBy', 'lastModifiedBy');
+	
+	// construct select columns list
+	$selectColumns = [];
+	if (!is_null($data) && !empty($data)) {
+		foreach ($defaultSelectColumns as $key) {
+			if (isset($data[$key])) {
+				if ($data[$key] == true) {
+					$selectColumns[] = $key; // append the select columns
+				}
+			}
+		}
+	}
+	$selectColumnsString = join(',', $selectColumns);
+
+	$year = 2018;
```

Comment:
Really not sure how I feel about this. I think if for some reason the year isn't set, then we should outright reject the creation of the appointment. The most likely way that this would happen is if someone manually sent the AJAX request (which we don't wont), and if someone is doing that, then we probably don't want to make the appointment.

Imagine it is 2020, and someone creates an appointment without a year. Then we end up with an entry for an appointment back in 2018, which invalidates all the 2018 data. I would be partial to removing this and instead rejecting the creation if the year is not set.

Reply:
This is for retrieving the shift times in order to determine when to make appointments. I simply set it to have a default year that was valid. I think it would make more sense to have the default be ```$year = date("Y")```. The alternatives when an invalid year is sent in would be to either have no shifts be returned or have every shift be returned.

Reply:
Oh yah I forgot what file we were in. I think returning no shifts is the way to go with an invalid year

Reply:
> `$year = date("Y")`
> have every shift be returned

Either of the above are completely valid in my opinion.